warp/yul/DeadcodeEliminator.py

```diff
@@ -0,0 +1,171 @@
+        return (live_vars, node)
+
+    def visit_function_definition(
+        self, node: ast.FunctionDefinition
+    ) -> (LiveVars, ast.FunctionDefinition):
+        # We need to ignore the generated getters from kudu since their
+        # body is only instantiated in ToCairoVisitor
+        if node.name.startswith("getter"):
+            return (set(), node)
+        (live_vars, body) = self.visit(
+            node.body, return_vars=set([n.name for n in node.return_variables])
+        )
+        live_vars = live_vars - set([arg.name for arg in node.parameters])
+        # The warp ifs are the only place where the if branch has a return in each branch
+        # if node.name.startswith("__warp_if"):
+        #     live_vars -= set([arg.name for arg in node.return_variables])
+        return (
+            live_vars,
+            ast.FunctionDefinition(
+                name=node.name,
+                parameters=node.parameters,
+                return_variables=node.return_variables,
+                body=body,
+            ),
+        )
+
+    def visit_if(
+        self, node: ast.If, live_vars: Set[str] = None, return_vars: Set[str] = None
+    ) -> Tuple[LiveVars, ast.If]:
+        if any(
+            [
+                isinstance(statement, ast.Leave)
+                for statement in node.body.statements
+                + (node.else_body.statements if node.else_body else ())
+            ]
+        ):
+            live_vars = return_vars | live_vars
+
+        (live_vars_cond, cond) = self.visit(node.condition)
+        (live_vars_body, body) = self.visit_block(node.body, live_vars)
+        (live_vars_else, else_body) = (
+            self.visit_block(node.else_body, live_vars)
+            if node.else_body
+            else (set(), None)
+        )
+        return (
+            live_vars_cond | live_vars_body | live_vars_else,
+            ast.If(condition=cond, body=body, else_body=else_body),
+        )
+
+    def visit_switch(self, node: ast.Switch) -> Tuple[LiveVars, ast.Switch]:
+        # TODO: Implement these so we can run DeadCodeEliminator at any stage
+        assert False, "There should be no switches, run SwitchToIfVisitor first"
+
+    def visit_forloop(self, node: ast.ForLoop) -> Tuple[LiveVars, ast.ForLoop]:
+        # TODO: Implement these so we can run DeadCodeEliminator at any stage
+        assert False, "There should be no for loops, run ForLoopIliminator first"
+
+    def visit_break(self, node: ast.Break) -> Tuple[LiveVars, ast.Break]:
+        # TODO: Implement these so we can run DeadCodeEliminator at any stage
+        assert False, "There should be no breaks, run ForLoopIliminator first"
+        return (set(), node)
+
+    def visit_continue(self, node: ast.Continue) -> Tuple[LiveVars, ast.Continue]:
+        # TODO: Implement these so we can run DeadCodeEliminator at any stage
+        assert False, "There should be no continues, run ForLoopIliminator first"
+        return (set(), node)
```

Comment:
Its okay to assume that you have a simplified tree. It's a good idea to write in the docs which visitors should go before this one though

Reply:
Yes, but in this case we could usefully add the eliminator at any stage. I'll update the docs